README
NAME
    match::simple - simplified clone of smartmatch operator

SYNOPSIS
       use v5.10;
       use match::simple;
   
       if ($this |M| $that)
       {
          say "$this matches $that";
       }

DESCRIPTION
    match::simple provides a simple match operator `|M|` that acts like a sane
    subset of the (as of Perl 5.18) deprecated smart match operator. Unlike
    smart match, the behaviour of the match is determined entirely by the
    operand on the right hand side.

    *   If the right hand side is `undef`, then there is only a match if the
        left hand side is also `undef`.

    *   If the right hand side is a non-reference, then the match is a simple
        string match.

    *   If the right hand side is a reference to a regexp, then the left hand
        is evaluated .

    *   If the right hand side is a code reference, then it is called in a
        boolean context with the left hand side being passed as an argument.

    *   If the right hand side is an object which provides a `MATCH` method,
        then it this is called as a method, with the left hand side being
        passed as an argument.

    *   If the right hand side is an object which overloads `~~`, then a true
        smart match is performed.

    *   If the right hand side is an arrayref, then the operator recurses into
        the array, with the match succeeding if the left hand side matches any
        array element.

    *   If any other value appears on the right hand side, the operator will
        croak.

    If you don't like the crazy Sub::Infix operator, you can alternatively
    export a more normal function:

       use v5.10;
       use match::simple qw(match);
   
       if (match($this, $that))
       {
          say "$this matches $that";
       }

    If you're making heavy use of this module, then this is probably your best
    option, as it runs significantly faster.

  XS Backend
    If you install match::simple::XS, a faster XS-based implementation will be
    used instead of the pure Perl functions. Depending on what sort of match
    you are doing, this is likely to be several times faster. In extreme
    cases, such as matching a string in an arrayref, it can be twenty-five
    times faster, or more. However, where $that is a single regexp, it's
    around 30% slower. Overall though, I think the performance improvement is
    worthwhile.

    If you want to take advantage of this speed up, use the `match` function
    rather than the `|M|` operator. Otherwise all your gains will be lost to
    the slow implementation of operator overloading.

    The constant `match::simple::IMPLEMENTATION` tells you which backend is
    currently in use.

  Environment
    Setting the `MATCH_SIMPLE_IMPLEMENTATION` environment variable to "PP"
    encourages match::simple to use the pure Perl backend.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=match-simple>.

SEE ALSO
    match::smart.

    This module uses Exporter::Tiny.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013-2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
match-simple
============

Created:      2013-05-19
Home page:    <https://metacpan.org/release/match-simple>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=match-simple>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.009	2014-07-02

 - Backport to Perl 5.6.
 - Don't load Sub::Infix unless needed.

0.008	2014-03-03

 [ Packaging ]
 - Bump recommended version of match::simple::XS to 0.001.

 [ Other ]
 - Don't use match::simple::XS if it's too old, or is unstable.

0.007_03	2014-03-02

 [ Test Suite ]
 - Don't do `no overloading` on Perl < 5.10.1 in test cases.

0.007_02	2014-03-02

 [ Test Suite ]
 - Improved match::simple test cases.

0.007_01	2014-02-21

 - Use match::simple::XS if it's installed.

0.006	2013-09-29

 [ Packaging ]
 - Fix dependencies.

0.005	2013-09-28

 [ Packaging ]
 - use Dist::Inkt

 [ Other ]
 - Updated: use Exporter::Tiny

0.004	2013-05-20

 [ Bug Fixes ]
 - Further fix smartmatch test cases on archaic Perl.

0.003	2013-05-20

 [ Bug Fixes ]
 - Fix smartmatch test cases on archaic Perl.

0.002	2013-05-20

 [ Bug Fixes ]
 - Better parity with Perl's smartmatch operator.
 - Proper checks for overloaded ~~.

 [ Packaging ]
 - Bundle tests based on Perl's tests for the smartmatch operator.

0.001	2013-05-19	Initial release


lib/match/smart.pm
package match::smart;

use 5.008001;
use strict;
use warnings;

use B qw();
use List::MoreUtils qw(any all);
use Scalar::Util qw(blessed looks_like_number refaddr);
use Sub::Infix qw(infix);

BEGIN {
	$match::smart::AUTHORITY = 'cpan:TOBYINK';
	$match::smart::VERSION   = '0.004';
}

use base "Exporter::TypeTiny";
our @EXPORT    = qw( M );
our @EXPORT_OK = qw( match );

sub match
{
	no warnings qw(uninitialized numeric);
	
	my ($a, $b, $seen) = @_;
	
	return(!defined $a)                    if !defined($b);
	return !!$b->check($a)                 if blessed($b) && $b->isa("Type::Tiny");
	return !!$b->MATCH($a, 1)              if blessed($b) && $b->can("MATCH");
	return eval 'no warnings; !!($a~~$b)'  if blessed($b) && $] >= 5.010 && do { require overload; overload::Method($b, "~~") };
	
	if (blessed($b) and not $b->isa("Regexp"))
	{
		require Carp;
		Carp::croak("Smart matching a non-overloaded object breaks encapsulation");
	}
	
	$seen ||= {};
	return refaddr($a)==refaddr($b) if $seen->{refaddr($b)}++;
	
	if (ref($b) eq q(ARRAY))
	{
		if (ref($a) eq q(ARRAY))
		{
			return !!0 unless @$a == @$b;
			for my $i (0 .. $#$a)
			{
				return !!0 unless match($a->[$i], $b->[$i], $seen);
			}
			return !!1;
		}
		
		return any { exists $a->{$_} } @$b  if ref($a) eq q(HASH);
		return any { $_ =~ $a } @$b         if ref($a) eq q(Regexp);
		return any { !defined($_) } @$b     if !defined($a);
		return any { match($a, $_) } @$b;
	}
	
	if (ref($b) eq q(HASH))
	{
		return match([sort map "$_", keys %$a], [sort map "$_", keys %$b])
			if ref($a) eq q(HASH);
		
		return any { exists $b->{$_} } @$a  if ref($a) eq q(ARRAY);
		return any { $_ =~ $a } keys %$b    if ref($a) eq q(Regexp);
		return !!0                          if !defined($a);
		return exists $b->{$a};
	}
	
	if (ref($b) eq q(CODE))
	{
		return all { !!$b->($_) } @$a       if ref($a) eq q(ARRAY);
		return all { !!$b->($_) } keys %$a  if ref($a) eq q(HASH);
		return $b->($a);
	}
	
	if (ref($b) eq q(Regexp))
	{
		return any { $_ =~ $b } @$a       if ref($a) eq q(ARRAY);
		return any { $_ =~ $b } keys %$a  if ref($a) eq q(HASH);
		return $a =~ $b;
	}
	
	return !!$a->check($b)                 if blessed($a) && $a->isa("Type::Tiny");
	return !!$a->MATCH($b)                 if blessed($a) && $a->can("MATCH");
	return eval 'no warnings; !!($a~~$b)'  if blessed($a) && $] >= 5.010 && do { require overload; overload::Method($a, "~~") };
	return !defined($b)                    if !defined($a);
	return $a == $b                        if _is_number($b);
	return $a == $b                        if _is_number($a) && looks_like_number($b);
	
	return $a eq $b;
}

sub _is_number
{
	my $value = shift;
	return if ref $value;
	my $flags = B::svref_2object(\$value)->FLAGS;
	$flags & ( B::SVp_IOK | B::SVp_NOK ) and !( $flags & B::SVp_POK );
}

*M = &infix(\&match);

__END__

=pod

=encoding utf-8

=for stopwords smartmatch recurses

=head1 NAME

match::smart - clone of smartmatch operator

=head1 SYNOPSIS

   use v5.10;
   use match::smart;
   
   if ($this |M| $that)
   {
      say "$this matches $that";
   }

=head1 DESCRIPTION

match::smart provides a match operator C<< |M| >> that acts like more or
less identically to the (as of Perl 5.18) deprecated smart match operator.

If you don't like the crazy L<Sub::Infix> operator, you can alternatively
export a more normal function:

   use v5.10;
   use match::smart qw(match);
   
   if (match($this, $that))
   {
      say "$this matches $that";
   }

=head2 Differences with ~~

There were major changes to smart match between 5.10.0 and 5.10.1. This
module attempts to emulate the behaviour of the operator in more recent
versions of Perl. In particular, 5.18.0 (minus the warnings). Divergences
not noted below should be considered bugs.

While the real smart match operator implicitly takes references to operands
that are hashes or arrays, match::smart's operator does not.

   @foo ~~ %bar       # means: \@foo ~~ \%bar
   @foo |M| %bar      # means: scalar(@foo) |M| scalar(%bar)

If you want the C<< \@foo ~~ \%bar >> behaviour, you need to add the
backslashes yourself:

   \@foo |M| \%bar

Similarly:

   "foo" ~~  /foo/    # works
   "foo" |M| /foo/    # no worky!
   "foo" |M| qr/foo/  # do this instead

match::smart treats the C<MATCH> method on blessed objects (if it exists)
like an overloaded C<< ~~ >>. This is for compatibility with L<match::simple>,
and for compatibility with pre-5.10 Perls that don't allow overloading
C<< ~~ >>.

=begin trustme

=item M

=item match

=end trustme

=head1 BUGS

Please report any bugs to
L<http://rt.cpan.org/Dist/Display.html?Queue=match-simple>.

=head1 SEE ALSO

L<match::simple>.

This module uses L<Exporter::TypeTiny>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT AND LICENCE

This software is copyright (c) 2013 by Toby Inkster.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 DISCLAIMER OF WARRANTIES

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



lib/match/simple.pm
package match::simple;

use 5.008001;
use strict;
use warnings;

use List::MoreUtils qw(any);
use Scalar::Util qw(blessed);
use Sub::Infix qw(infix);

BEGIN {
	$match::simple::AUTHORITY = 'cpan:TOBYINK';
	$match::simple::VERSION   = '0.004';
}

use base "Exporter::TypeTiny";
our @EXPORT    = qw( M );
our @EXPORT_OK = qw( match );

sub match
{
	no warnings qw(uninitialized numeric);
	
	my ($a, $b) = @_;
	
	return(!defined $a)                    if !defined($b);
	return($a eq $b)                       if !ref($b);
	return($a =~ $b)                       if ref($b) eq q(Regexp);
	return do{ local $_ = $a; !!$b->($a) } if ref($b) eq q(CODE);
	return any { match($a, $_) } @$b       if ref($b) eq q(ARRAY);
	return !!$b->check($a)                 if blessed($b) && $b->isa("Type::Tiny");
	return !!$b->MATCH($a, 1)              if blessed($b) && $b->can("MATCH");
	return eval 'no warnings; !!($a~~$b)'  if blessed($b) && $] >= 5.010 && do { require overload; overload::Method($b, "~~") };
	
	require Carp;
	Carp::croak("match::simple cannot match anything against: $b");
}

*M = &infix(\&match);

__END__

=pod

=encoding utf-8

=for stopwords smartmatch recurses

=head1 NAME

match::simple - simplified clone of smartmatch operator

=head1 SYNOPSIS

   use v5.10;
   use match::simple;
   
   if ($this |M| $that)
   {
      say "$this matches $that";
   }

=head1 DESCRIPTION

match::simple provides a simple match operator C<< |M| >> that acts like
a sane subset of the (as of Perl 5.18) deprecated smart match operator.
Unlike smart match, the behaviour of the match is determined entirely by
the operand on the right hand side.

=over

=item *

If the right hand side is C<undef>, then there is only a match if the left
hand side is also C<undef>.

=item *

If the right hand side is a non-reference, then the match is a simple string
match.

=item *

If the right hand side is a reference to a regexp, then the left hand is
evaluated .

=item *

If the right hand side is a code reference, then it is called in a boolean
context with the left hand side being passed as an argument.

=item *

If the right hand side is an object which provides a C<MATCH> method, then
it this is called as a method, with the left hand side being passed as an
argument.

=item *

If the right hand side is an object which overloads C<~~>, then a true
smart match is performed.

=item *

If the right hand side is an arrayref, then the operator recurses into the
array, with the match succeeding if the left hand side matches any array
element.

=item *

If any other value appears on the right hand side, the operator will croak.

=back

If you don't like the crazy L<Sub::Infix> operator, you can alternatively
export a more normal function:

   use v5.10;
   use match::simple qw(match);
   
   if (match($this, $that))
   {
      say "$this matches $that";
   }

=begin trustme

=item M

=item match

=end trustme

=head1 BUGS

Please report any bugs to
L<http://rt.cpan.org/Dist/Display.html?Queue=match-simple>.

=head1 SEE ALSO

L<match::smart>.

This module uses L<Exporter::TypeTiny>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT AND LICENCE

This software is copyright (c) 2013 by Toby Inkster.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 DISCLAIMER OF WARRANTIES

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "simplified clone of smartmatch operator",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.016, CPAN::Meta::Converter version 2.140640",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "match-simple",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        runtime => {
                          recommends => { "match::simple::XS" => 0.001 },
                          requires   => {
                                          "Exporter::Tiny" => 0.026,
                                          "List::Util"     => 1.33,
                                          "perl"           => 5.006001,
                                          "Scalar::Util"   => 0,
                                          "Sub::Infix"     => 0.004,
                                        },
                        },
                        test => { requires => { "Test::Fatal" => 0, "Test::More" => 0.96 } },
                      },
  "provides"       => {
                        "match::simple" => { file => "lib/match/simple.pm", version => 0.009 },
                        "match::smart"  => { file => "lib/match/smart.pm", version => 0.009 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => {
                                          web => "http://rt.cpan.org/Dist/Display.html?Queue=match-simple",
                                        },
                        homepage     => "https://metacpan.org/release/match-simple",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-match-simple.git",
                                          web  => "https://github.com/tobyink/p5-match-simple",
                                        },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/match-simple/project",
                      },
  "version"        => 0.009,
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

